cranelift/isle/isle/src/serialize.rs

```diff
@@ -0,0 +1,523 @@
+            self.candidates.extend(
+                // A binding site will have at most one of these kinds of constraint, and many have
+                // none. But `best_single_binding` has to check all candidates anyway, so let it
+                // figure out which (if any) of these are applicable. It will only check false
+                // candidates once on any partition, removing them from this list immediately.
+                [
+                    MatchKind::Constraint(source),
+                    MatchKind::Iterator(source),
+                    // Obviously this binding site equals itself, so this checks whether it
+                    // participates in any equality constraints at all.
+                    MatchKind::Equal(source, source),
+                ]
+                .into_iter()
+                .map(|kind| Candidate {
+                    // count will be filled in later by `best_single_binding`
+                    count: 0,
+                    state,
+                    kind,
+                }),
+            );
+        }
+
+        *old = state;
+    }
+
+    fn sort(mut self, mut order: &mut [usize]) -> Block {
+        while !order.is_empty() {
+            if let Some(best) = self.best_single_binding(order) {
+                let partition_point = split_on(&self.rules, order, best);
+                debug_assert!(partition_point > 0);
+                let (this, rest) = order.split_at_mut(partition_point);
+                order = rest;
+
+                let check = match best {
+                    MatchKind::Constraint(source) => {
+                        self.use_expr(source);
+                        self.add_bindings();
+                        let mut arms = Vec::new();
+
+                        let get_constraint =
+                            |idx: usize| self.rules.rules[idx].get_constraint(source).unwrap();
+                        this.sort_unstable_by_key(|&idx| get_constraint(idx));
+                        for g in group_by_mut(this, |&a, &b| get_constraint(a) == get_constraint(b))
+                        {
+                            let mut child = self.new_block();
+                            // Applying a constraint moves the discriminant from Emitted to Matched, but
+                            // only within the constraint's match arm; later fallthrough cases may need to
+                            // match this discriminant again.
+                            child.set_ready(source, Readiness::Matched);
+
+                            let constraint = get_constraint(g[0]);
+                            let bindings = Vec::from_iter(
+                                constraint
+                                    .bindings_for(source)
+                                    .into_iter()
+                                    .map(|b| child.rules.find_binding(&b)),
+                            );
+
+                            let mut changed = false;
```

Comment:
`changed` here needs to consider whether `set_ready` above (line 225) updated the readiness of `source` as well